ArmPkg/Library/ArmFfaLib/ArmFfaDxeLib.c

```diff
@@ -0,0 +1,156 @@
+/** @file
+  Arm Ffa library code for Dxe Driver
+
+  Copyright (c) 2024, Arm Limited. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+   @par Glossary:
+     - FF-A - Firmware Framework for Arm A-profile
+
+   @par Reference(s):
+     - Arm Firmware Framework for Arm A-Profile [https://developer.arm.com/documentation/den0077/latest]
+
+**/
+
+#include <Uefi.h>
+#include <Pi/PiMultiPhase.h>
+
+#include <Library/ArmLib.h>
+#include <Library/ArmSmcLib.h>
+#include <Library/ArmFfaLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/PcdLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+#include <IndustryStandard/ArmFfaSvc.h>
+
+#include "ArmFfaCommon.h"
+#include "ArmFfaRxTxMap.h"
+
+STATIC EFI_EVENT  mFfaExitBootServiceEvent;
+
+/**
+  Unmap RX/TX buffer on Exit Boot Service.
+
+  @param [in]   Event      Registered exit boot service event.
+  @param [in]   Context    Additional data.
+
+**/
+STATIC
+VOID
+EFIAPI
+ArmFfaLibExitBootServiceEvent (
+  IN EFI_EVENT  Event,
+  IN VOID       *Context
+  )
+{
+  ArmFfaLibRxTxUnmap ();
+}
+
+/**
+  ArmFfaLib Constructor.
+
+  @param [in]   ImageHandle      Image Handle
+  @param [in]   SystemTable      System Table
+
+  @retval EFI_SUCCESS            Success
+  @retval EFI_INVALID_PARAMETER  Invalid alignment of Rx/Tx buffer
+  @retval EFI_OUT_OF_RESOURCES   Out of memory
+  @retval Others                 Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaDxeLibConstructor (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS                 Status;
+  EFI_HOB_GUID_TYPE          *RxTxBufferHob;
+  ARM_FFA_RX_TX_BUFFER_INFO  *BufferInfo;
+
+  Status = ArmFfaLibCommonInit ();
+  if (EFI_ERROR (Status)) {
+    if (Status == EFI_UNSUPPORTED) {
+      /*
+       * EFI_UNSUPPORTED return from ArmFfaLibCommonInit() means
+       * FF-A interface doesn't support.
+       * However, It doesn't make failure of loading driver/library instance
+       * (i.e) Driver can be written with SpmMm if FF-A doesn't support.

```

Comment:
Please explain that the MM_COMMUNICATE Dxe uses this library as well as SpmMM. So if FF-A is not supported the MM_COMMUNICATE Dxe falls back to SpmMM.